[Clock] Documentation for new Clock component (#17435)
symfony/symfony-docs: components/clock.rst

            $expirationChecker = new ExpirationChecker($clock);
            $validUntil = new DateTimeImmutable('2022-11-16 15:25:00');

            // $validUntil is in the future, so it is not expired
            static::assertFalse($expirationChecker->isExpired($validUntil));

            // Clock sleeps for 10 minutes, so now is '2022-11-16 15:30:00'
            $clock->sleep(600); // Instantly changes time as if we waited for 10 minutes (600secs)

            // modify the clock, accepts all formats supported by DateTimeImmutable::modify()
            static::assertTrue($expirationChecker->isExpired($validUntil));

            $clock->modify('2022-11-16 15:00:00');

            // $validUntil is in the future again, so it is no longer expired
            static::assertFalse($expirationChecker->isExpired($validUntil));
        }
    }

Monotonic Clock
---------------

The ``MonotonicClock`` allows you to implement a precise stopwatch, depending on the system up to
nanosecond precision. It can be used to measure the elapsed time between 2 calls without being
affected by inconsistencies sometimes introduced by the system clock, e.g. by updating it. Instead,

> it consistently increases time, making it especially useful for measuring performance consistently.

Reviewer: consistently ist used twice here

.. index::
   single: Clock
   single: Components; Clock

.. _`clock-component`:

The Clock Component
===================

> 

Reviewer: `versionadded` is missing

The Clock component decouples applications from the system clock. This improves
testability of time-sensitive logic.

> The component provides a ``ClockInterface`` with 3 different clock

Reviewer: ```suggestion
The component provides a ``ClockInterface`` with the following clock
```

implementations for different use cases.

* ``NativeClock`` provides a way to interact with the system clock.
* ``MockClock`` is commonly used in tests as a replacement for the
  ``NativeClock`` to be able to freeze and change the current time using either
  ``sleep()`` or ``modify()``.
* ``MonotonicClock`` relies on ``hrtime()`` and provides a high resolution,
  monotonic clock, when you need a precise stopwatch.

Installation
------------

.. code-block:: terminal

    $ composer require symfony/clock

.. include:: /components/require_autoload.rst.inc

Usage
-----

A clock service replaces creating a new ``DateTime`` or

> ``DateTimeImmutable``-object for the current time. Instead, you inject the

Reviewer: ```suggestion
``DateTimeImmutable`` object for the current time. Instead, you inject the
```

``ClockInterface`` and call ``now()``. By default, your application will likely
use a ``NativeClock``, which always returns the current system time. In tests it is replaced by a ``MockClock``. The
``MockClock`` is instantiated with a time and does not move forward on its own, meaning when you create an
instance with a time, it is always kept, unless you change it using
either ``sleep()`` or ``modify()``. This gives you full control over what your
code assumes is the current time.

The following example introduces a service utilizing the clock component to

> determine the current time.

Reviewer: ```suggestion
determine the current time::
```

    use Symfony\Component\Clock\ClockInterface;

    class ExpirationChecker
    {
        public function __construct(
            private readonly ClockInterface $clock
        ) {}

        public function isExpired(DateTimeInterface $time): bool
        {
            return $this->clock->now() > $time;
        }
    }

> 

Reviewer: Could have its own headline IMO

Author: I'm sorry, I don't understand? What headline do you have in mind?

Reviewer: ```suggestion
Testing
-------
```
👍

When writing a test for this service, you can check both cases where something

> is expired or not, by modifying the clock's time.

Reviewer: ```suggestion
is expired or not, by modifying the clock's time::
```

>     use Symfony\Component\Clock\MockClock;

Reviewer: Lets add a `use` for `TestCase`

    class ExpirationCheckerTest extends TestCase
    {
        public function testIsExpired(): void
        {
            $clock = new MockClock('2022-11-16 15:20:00');
            $expirationChecker = new ExpirationChecker($clock);
            $time = new DateTimeImmutable('2022-11-16 15:25:00');

            // $time is in the future, so it is not expired
            static::assertFalse($expirationChecker->isExpired($time));

>             // Clock sleeps for 10min, so now is '2022-11-16 15:30:00'

Reviewer: ```suggestion
            // Clock sleeps for 10 minutes, so now is '2022-11-16 15:30:00'
```

>             $clock->sleep(600); // Instantly changes time as if we waited for 10min (600secs)

Reviewer: ```suggestion
            $clock->sleep(600); // Instantly changes time as if we waited for 10 minutes (600secs)
```

.. index::
   single: Clock
   single: Components; Clock

.. _`clock-component`:

The Clock Component
===================

.. versionadded:: 6.2

    The Clock component was introduced in Symfony 6.2

> The Clock component decouples applications from the system clock. This improves
> testability of time-sensitive logic.

Reviewer 2: ```suggestion
The Clock component decouples applications from the system clock. This allows
you to fix time to improve testability of time-sensitive logic.
```

just a suggestion, but I think we should clarify a bit on the *why* about improving testability

The component provides a ``ClockInterface`` with the following implementations
for different use cases:

> * ``NativeClock`` provides a way to interact with the system clock;
> * ``MockClock`` is commonly used in tests as a replacement for the
>   ``NativeClock`` to be able to freeze and change the current time using either
>   ``sleep()`` or ``modify()``;
> * ``MonotonicClock`` relies on ``hrtime()`` and provides a high resolution,
>   monotonic clock, when you need a precise stopwatch.

Reviewer 2: ```suggestion
:class:`Symfony\\Component\\Clock\\NativeClock`
    Provides a way to interact with the system clock, this is the same as doing
    ``new \DateTimeImmutable()``.
:class:`Symfony\\Component\\Clock\\MockClock`
    Commonly used in tests as a replacement for the ``NativeClock`` to be able
    to freeze and change the current time using either ``sleep()`` or ``modify()``.
:class:`Symfony\\Component\\Clock\\MonotonicClock``
    Relies on ``hrtime()`` and provides a high resolution, monotonic clock,
    when you need a precise stopwatch.
```

definition lists for the win :)

Installation
------------

.. code-block:: terminal

    $ composer require symfony/clock

.. include:: /components/require_autoload.rst.inc

Usage
-----

A clock service replaces creating a new ``DateTime`` or
``DateTimeImmutable`` object for the current time. Instead, you inject the
``ClockInterface`` and call ``now()``. By default, your application will likely
use a ``NativeClock``, which always returns the current system time. In tests it is replaced with a ``MockClock``. The
The ``MockClock`` is instantiated with a time and does not move forward on its own. The time is
fixed until ``sleep()`` or ``modify()`` are called. This gives you full control over what your code

> assumes is the current time.

Reviewer 2: I think we should move the mock clock information the testing section below (and rename it to "Testing using MockClock"?). Then we can also explain that sleep doesn't actually sleep using the mock clock.

        ) {}

        public function isExpired(DateTimeInterface $validUntil): bool
        {
            return $this->clock->now() > $validUntil;
        }
    }

Testing
-------

When writing a test for this service, you can check both cases where something
is expired or not, by modifying the clock's time::

    use PHPUnit\Framework\TestCase;
    use Symfony\Component\Clock\MockClock;

    class ExpirationCheckerTest extends TestCase
    {
        public function testIsExpired(): void
        {
            $clock = new MockClock('2022-11-16 15:20:00');
            $expirationChecker = new ExpirationChecker($clock);
            $validUntil = new DateTimeImmutable('2022-11-16 15:25:00');

            // $validUntil is in the future, so it is not expired
            static::assertFalse($expirationChecker->isExpired($validUntil));

            // Clock sleeps for 10 minutes, so now is '2022-11-16 15:30:00'
            $clock->sleep(600); // Instantly changes time as if we waited for 10 minutes (600secs)

            // $validUntil is in the past after advancing the clock, so it is expired
            static::assertTrue($expirationChecker->isExpired($validUntil));

>             $clock->modify('2022-11-16 15:00:00');

Reviewer 2: ```suggestion
            // modify the clock, accepts all formats supported by DateTimeImmutable::modify()
            $clock->modify('2022-11-16 15:00:00');
```

.. index::
   single: Clock
   single: Components; Clock

> .. _`clock-component`:

Reviewer 2: ```suggestion
```

(this system is a custom anchor - but we don't use it, so let's remove it)

The Clock Component
===================

.. versionadded:: 6.2

    The Clock component was introduced in Symfony 6.2

The Clock component decouples applications from the system clock. This improves
testability of time-sensitive logic.

The component provides a ``ClockInterface`` with the following implementations

> for different use cases.

Reviewer 3: ```suggestion
for different use cases:
```

> * ``NativeClock`` provides a way to interact with the system clock.

Reviewer 3: ```suggestion
* ``NativeClock`` provides a way to interact with the system clock;
```

* ``MockClock`` is commonly used in tests as a replacement for the
  ``NativeClock`` to be able to freeze and change the current time using either

>   ``sleep()`` or ``modify()``.

Reviewer 3: ```suggestion
  ``sleep()`` or ``modify()``;
```

* ``MonotonicClock`` relies on ``hrtime()`` and provides a high resolution,
  monotonic clock, when you need a precise stopwatch.

Installation
------------

.. code-block:: terminal

    $ composer require symfony/clock

.. include:: /components/require_autoload.rst.inc

Usage
-----

A clock service replaces creating a new ``DateTime`` or
``DateTimeImmutable`` object for the current time. Instead, you inject the
``ClockInterface`` and call ``now()``. By default, your application will likely

> use a ``NativeClock``, which always returns the current system time. In tests it is replaced by a ``MockClock``. The

Reviewer 3: replaced *with*?

``MockClock`` is instantiated with a time and does not move forward on its own, meaning when you create an
instance with a time, it is always kept, unless you change it using

> either ``sleep()`` or ``modify()``. This gives you full control over what your

Reviewer 3: The `MockClock` is instantiated with a time and does not move forward on its own. The time is fixed until `sleep()` or `modify()` are called.

code assumes is the current time.

> The following example introduces a service utilizing the clock component to

Reviewer 3: ```suggestion
The following example introduces a service utilizing the Clock component to
```

determine the current time::

    use Symfony\Component\Clock\ClockInterface;

    class ExpirationChecker
    {
        public function __construct(
            private readonly ClockInterface $clock
        ) {}

>         public function isExpired(DateTimeInterface $time): bool

Reviewer 3: Maybe rename `$time` to `$validUntil`?
